src/main/java/teammates/logic/core/FeedbackResponsesLogic.java

```diff
@@ -421,11 +426,32 @@ public void updateFeedbackResponse(
         
         // Create a copy.
         FeedbackResponseAttributes newResponse = new FeedbackResponseAttributes(updatedResponse);
-        
         FeedbackResponseAttributes oldResponse = new FeedbackResponseAttributes(oldResponseEntity);
 
         // Copy values that cannot be changed to defensively avoid invalid
         // parameters.
+        copyFromOldToNew(newResponse, oldResponse);
+    
+        if (newResponse.recipient.equals(oldResponse.recipient)
+                && newResponse.giver.equals(oldResponse.giver)) {
+            try {
+                frDb.updateFeedbackResponseOptimized(newResponse, oldResponseEntity);
+            } catch (EntityDoesNotExistException e) {
+                Assumption.fail();
+            }
+        } else {
+            // Recreate response to prevent possible future id conflict.
+            recreateResponse(newResponse, oldResponse);
+        }
+    }
+
+    /**
+     * Copy values that cannot be changed to defensively avoid invalid parameters.
+     * @param newResponse  values are copied from oldResponse
+     * @param oldResponse  values are copied to newResponse
+     */
+    private void copyFromOldToNew(FeedbackResponseAttributes newResponse,
```

Comment:
Still think the name of the method could be improved - perhaps `copyFixedValuesFromOldToNew`?